testrpc/wait.go

```diff
@@ -20,8 +20,8 @@ func WaitForLeader(t *testing.T, rpc rpcFn, dc string) {
 		if !out.QueryMeta.KnownLeader {
 			r.Fatalf("No leader")
 		}
-		if out.Index == 0 {
-			r.Fatalf("Consul index is 0")
+		if out.Index < 2 {
+			r.Fatalf("Consul index should be at least 2")
```

Comment:
I don't understand how this fails _less_ often since the new condition matches strictly _more_ states than the old one?

i.e. if used to only fail of `Index == 0` now it will still fail for `Index == 0` but also `Index == 1`. This might be correct but how does it make tests _less_ flaky?

Reply:
This code checks until the Server is ready => thus with Index 1 => it seems a change did occur regarding indexes and did return index == 1 while the Server was actually not ready => tests relying on WaitForLeader() to wait for leader were unstable

Reply:
Ah right, the diff hides that this is in a retry loop, so being _stricter_ on failure here actually allows it to retry for longer until it really is ready.

This seems legit and I think my recent blocking change was the root cause of this getting worse so this is a great find.

Reply:
@banks DONE, I reverted the removal of t.Parallel() I did remove, but I expect it won't work, let's see what Travis says.

t.Parallel() really does not work well when using  ̀NewTestAgent` probably because it interfere with other tests using it at the same time, probably because of network (and all having the same DC). A fix would be to store in the test context a kind of shared ID (with network isolation if it is the root cause) so tests would not interact one with the other. But I did not spent enough time on this, this is pure hypothesis on my side

Reply:
> t.Parallel() really does not work well when using ̀NewTestAgent` probably because it interfere with other tests using it at the same time, probably because of network (and all having the same DC).

I might be wrong but t.Parallel only has any affect if `--parallel` is greater than one. Which is why I suggested not changing it since we can control that externally.

`NewTestAgent` is written specifically to be isolated - it picks random unused ports and state directories every time so it should be completely safe to run those test in parallel.

In practice each instance takes a non-trivial amount of CPU time so on CPU-constrained environments or where there is a lot of test parallelism, they run slower and so cause flakyness. but the _should_ be perfectly safe to run in parallel with enough CPU power and that is already controlled for in the CI scripts.

Thanks for making the change.

I'm going to let @MagnumOpus21 and @freddygv take this on as they have a better idea how well changes here will interact with their research and experiments.

I'd strongly advise keeping changes in small PRs rather than bundling lots of different test changes at once every time as it becomes impossible to know which change actually had an effect on flakes later on (it might have seemed better on a few runs locally but it's hard to know that later).

Reply:
@banks @MagnumOpus21 Ok, I think I will have most of tests fixed.

Sorry, I know the PR is very long, I will rewrite it to single tests if you want.

Most of the errors are actually due to `NewTestAgent()` being too slow to start on a heavily loaded machine, so the big fix - for most of the tests - is to use `WaitForLeader` which ensures the node is ready *before* the test really starts.

I thought using `WaitForLeader()` within `NewTestAgent()`, unfortunately some tests rely on the Agent not starting properly, so this is not possible for now.

I did slow down my laptop by running expensive tests and running at the same time:

```
while true;
do
  go clean -testcache
  make test-ci
done
```

and getting all failing tests. Now, I almost have no errors but a few similar to those https://travis-ci.org/hashicorp/consul/jobs/411685497 `err: Unexpected response code: 500 (Missing check 'serfHealth' registration)` sometimes.

By retrying 1 or 2 times on Travis, it should be fine most of the time now